src/deploy/functions/build.ts

```diff
@@ -0,0 +1,397 @@
+import * as backend from "./backend";
+import * as proto from "../../gcp/proto";
+import { FirebaseError } from "../../error";
+
+/* The union of a customer-controlled deployment and potentially deploy-time defined parameters */
+export interface Build {
+  requiredAPIs: RequiredApi[];
+  endpoints: Record<string, Endpoint>;
+  params: Param[];
+}
+
+interface RequiredApi {
+  // The API that should be enabled. For Google APIs, this should be a googleapis.com subdomain
+  // (e.g. vision.googleapis.com)
+  api: string;
+
+  // A reason why this codebase requires this API.
+  // Will be considered required for all Extensions codebases. Considered optional for Functions
+  // codebases.
+  reason?: string;
+}
+
+// Defining a StringParam with { param: "FOO" } declares that "{{ params.FOO }}" is a valid
+// Expression<string> elsewhere.
+// Expression<number> is always an int. Float Params, list params, and secret params cannot be used in
+// expressions.
+// `Expression<Foo> == Expression<Foo>` is an Expression<boolean>
+// `Expression<boolean> ? Expression<T> : Expression<T>` is an Expression<T>
+type Expression<T extends string | number | boolean> = string;
+
+function resolveInt(from: number | Expression<number> | null): number {
+  if (from == null) {
+    return 0;
+  } else if (typeof from === "string") {
+    throw new FirebaseError("CEL evaluation of expression '" + from + "' not yet supported");
+  }
+  return from;
+}
+
+function resolveString(from: string | Expression<string> | null): string {
+  if (from == null) {
+    return "";
+  }
```

Comment:
nit: Maybe check for "{{ }}" and throw here too.

Reply:
Done.